[FIX] purchase, sale: missing product name in journal items

Megaaaaaa opened the pull request:
Description of the issue this commit addresses:

When creating an invoice from the sales or the purchase app, the product names are not picked and put in the name (label) of the invoice lines linked to it. This causes several issues among which the act that the pdf of the invoice only gives the description of the product but not its name.

---

Desired behavior after this commit is merged:

This commit makes sure the product's display_name is always shown in the invoice line's name so that it is rendered on the pdf.

---

Enterprise PR:
required for opw-4119507 to be properly fixed

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

robodoo commented:
[![Pull request status dashboard](https://mergebot.odoo.com/odoo/odoo/pull/177242.png)](https://mergebot.odoo.com/odoo/odoo/pull/177242)

Feyensv reviewed (commented):
I don't really understand when the product name is not in the SOL `name` :thinking:.
Could you add a test to cover this fix (would also help understand the issue) ?

Megaaaaaa commented:
> I don't really understand when the product name is not in the SOL `name` 🤔. Could you add a test to cover this fix (would also help understand the issue) ?

@Feyensv 

Well the invoice line's name is only ever the SOL's name so any time the description of a sale order line is modified, the invoice's name and the pdf of said invoice are also modified. To fix that, instead of taking the SOL name blindly when preparing the invoices values, I check (with the `redundant_product_and_description` variable) whether the name of the product is in the description and vice versa and if neither then the product name must be added to the "name" of the aml that is being prepared. 

I've added steps to reproduce in the commits and the PR message, those describe what is done in the video provided in the support ticket but you can do it without sale_subscription by just removing the name of the product from the description of a SOL before invoicing it.

Also i'll be pushing an updated version of the commit with some tests soon. Thank you for the feedback.

Megaaaaaa commented:
@Feyensv Just pushed some tests. :smile:

And to clarify a little because i re-read my first message and i thought that it lacked this: You're right, initally the SOL name alway contains the product name but as some processes modify it (like manually modifying it), the transition from the sale order to the invoice loses data. We can't disable the modification of the description or else of what use would it be but we can't either have invoices without the product's name. That's why the modification is done at the very start of the preparation of the invoices lines.

Feyensv commented:
Okay, so only a problem specific to 17.4 since in 17.5 we also use the widget in purchase and sale.
Kinda magic solution (and not very solid, what if the product name is modified ?), not sure it's worth fixing TBH.

IMHO, it also highlights the fact that removing the product names from the descriptions might be better someday (with another solution for the description on journal items ofc).

Megaaaaaa commented:
Well, no, the bug is also present in master. It has nothing to do with the product/label widget, it is related to the fact the the SOL name can be modified and not contain the product name.

Also, the solution is not magic imo, checking whether the description contains the product name or not is pretty straightforward but since i have developed the fix, i guess i have a better understanding of what happens.
In cases in which the product name is changed, just as always, it checks whether it is in the description or not to make sure it's on the invoice one and only one time.

And finally, this is definitely worth fixing, internal is currently using a workaround to have proper invoices. Look at the tests, the ticket and reproduce the steps, you'll understand how bad it is.

I agree with you that someday, journal items label is going to need its own field but atm, we need a fix, not an imp.

Megaaaaaa commented:
> Hello @Megaaaaaa , while working on bugfix ticket related to this, this PR will fix my ticket but I noticed that the JS method responsible on cleaning the label from product name, doesn't take into account translation so we have duplicates :
> 
> https://github.com/odoo/odoo/blob/3addf419ace500d6dde6afe5cc9da4ef3fe34682/addons/account/static/src/components/product_label_section_and_note_field/product_label_section_and_note_field.js#L167-L176
> 
> 
> ![image](https://private-user-images.githubusercontent.com/97512966/366041699-510e308f-b953-4680-99d4-87d839a2afa3.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjYwNDk4MDEsIm5iZiI6MTcyNjA0OTUwMSwicGF0aCI6Ii85NzUxMjk2Ni8zNjYwNDE2OTktNTEwZTMwOGYtYjk1My00NjgwLTk5ZDQtODdkODM5YTJhZmEzLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA5MTElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwOTExVDEwMTE0MVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTZhOGZiYWRjMDViNzBjYjQyNmRlMTRhN2E3MmY1NDQxN2QxNWYwNTFlYmQ3YWFhYzQwMTkxZmUxNDAxYjZjZGUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.IUKV6H7waTa4GTugUFNRBihvJX6GEXV720lqy0Nruc4)

Hello :wave: 

Thank you for noticing this, as it's not related to this fix, i'll have another PR to fix it. It is indeed not intended. I'll handle it when I have a little more time as I am currently not able to do it.

abla001 commented:
Hello @Megaaaaaa @Feyensv @arj-odoo, can we merge the fix before next week? We keep receiving tickets for this issue in bugfix, and some clients have been unable to issue invoices for subscriptions since the start of the month.

Megaaaaaa commented:
Hello @abla001 I've received a review from @aboo-odoo, i'll push the changes this morning and should go to merge this week :+1:

aboo-odoo reviewed (commented):
LGTM! @clbr-odoo You're turn to bully him

jco-odoo commented on addons/purchase/models/purchase_order_line.py:
```diff
@@ -558,7 +558,7 @@ def _prepare_account_move_line(self, move=False):
         res = {
             'display_type': self.display_type or 'product',
-            'name': self.name,
+            'name': self.name if not self.product_id.display_name or (self.product_id.display_name in self.name or self.name in self.product_id.display_name) else f"{self.product_id.display_name} {self.name}",
```
We can not make a helper method for this in accounting on account.move.line?  (maybe there is one already)

jco-odoo commented:
@robodoo r+

robodoo commented:
@Megaaaaaa @jco-odoo linked pull request(s) odoo/enterprise#68620 not ready. Linked PRs are not staged until all of them are ready.

jco-odoo commented:
@robodoo r+
